docs/src/DeveloperDocumentation/debugging.md

```diff
@@ -0,0 +1,209 @@
+# Debugging Oscar Code
+
+## Pitfalls: Mutable objects in Oscar code
+
+Suppose you are having the following difficulties. Your code is exhibiting
+inexplicable behaviour and values that should not be changing are changing in
+seemingly random locations. To get to the bottom of these kind of issues
+it is necessary to be familiar with mutable objects in Julia and some of the
+relevant conventions in place in Oscar. This section discusses these
+informal rules as well as some of the exceptions to these rules.
+
+In Julia, objects that can change after construction are declared with the
+`mutable struct` keywords and satisfy the `ismutable` predicate.
+These objects can be linked together into an arbitrary dependency graph, and
+a change to one object may therefore have unintended consequences on another
+object in the system.
+
+The simplest example is the creation of a polynomial ring. If we mutate the
+array of symbols used for printing, we have effectively changed the ring.
+
+```
+julia> v = [:x, :y, :z]; R = PolynomialRing(QQ, v)[1]
+Multivariate Polynomial Ring in x, y, z over Rational Field
+
+julia> v[3] = :w; R
+Multivariate Polynomial Ring in x, y, w over Rational Field
+```
+
+In this example, the modification of `v` is unexpected and may in fact corrupt
+the internal data structures used by the polynomial ring. As such, this
+modification of `v` has to be considered illegal. Upon creation of the array
+called `v`, we have full rights over the object and can mutate at will.
+However, after passing it to the function `PolynomialRing`, we have given up
+*owernership* of the array and are no longer free to modify it.
+
+General Oscar Principle (GOP):
```

Comment:
I am not sure I like being associated with the GOP, but I guess I must read this non-political 😂